README.md
# Websense/Forcepoint WCG Auto Backup Script
This is a script that was designed to run on Websense/Forcepoint appliances to create automatic backups of the WCG settings "WCG Snapshot" and FTP them to an FTP Server. Currently the script creates a tar.gz file of the backup to upload, in the future it will be changed to create the same directory structure that the proxy does for easier remote restores.

Recently there was a custom log added which can be viewed in the WCG UI. The log can be accessed via Configure > Logs > System and then select "snapshot_script.log" from the dropdown box. An example of the log:

``` 2016-04-25:07:41:29 - Backup wcghostname.tar.gz was created and transfered to the FTP Server! MD5 of file is: 760ec68b1203eebc7b1579f7965d2a67 ```

This script is provided without any warranty or support. Any issues feel free to post them here on Github so I can look at them.


###Installation:
To install copy download and move the config.ini and wcg_auto_snapshot.py files to a directory of your choice. Modify the config.ini with your values and then create a cron job with the following to run every Sunday night at midnight:

```00 00 * * * python /your/directory/wcg_auto_snapshot.py```

#####SCP Setup:
To use SCP you need to setup your keys ahead of time. To generate a new set of private keys on the Content Gateway machine the following command and follow the prompts, do not set a password or the script will fail:

```root@proxy# ssh-keygen -t rsa```

Then you need to export the key to your SCP server:

```root@proxy# ssh-copy-id user@server```

Once this is done you need to enable SCP in the config.ini by changing the value to "True"


###Planned Changes:
- Change from tar.gz to exact directory structure WCG uses for easy remote restores.
- (COMPLETED) Move the configuration values to a .ini file for easy setup.
- (COMPLETED) Migrate from bash to Python for the script.
- Create an installer.
- (Completed) Add SCP support with SSH keys for better security.


*Notes: This script was tested working and developed on a Websense/Forcepoint V1000G3 running version 8.0.1


CHANGELOG
Changelog:

April 25, 2016
   0.9  - Initial release canidate version.
   0.91 - Fixed an issue with FTP not uploading in binary mode.
   0.92 - Added MD5 value written to log so that verification can be done on FTP server
  
April 26, 2016
   0.93 - Rewrote the script using Python 2.x, will leave the bash script for a bit until fully stable

May 05, 2016
   0.93a - Fixed a few bugs with removal of tar.gz using python script
   0.94 - Added support for SCP witout having to install additional python modules


old_files/wcg_auto_snapshot.sh
#!/bin/bash
#=========================================================================================
#title               :wcg_auto_snapshot.sh
#description         :This script is called by a cron job to take snapshot and FTP them.
#author              :Alan Matson "alan (at) echotek (dot) us"
#date                :2016-04-20
#version             :0.91    
#notes               :script is called via crontab "00 00 * * * bash /root/wcg_auto_snapshot.sh"
#websense_version    :8.0.1 (Running CentOS 6.3 Kernel 3.6.3)
#license             :GPLv3
#changelog           :https://github.com/ajmatson/Websense-WCG-Auto-Backup-Script/blob/master/CHANGELOG
#=========================================================================================

#Change the values for your FTP Server and local directory
ftpserverip="1.1.1.1"
username="USERNAME"
password="PASSWORD"
remotedir="/WebsenseBackups/WCG"
localdir="/opt/WCG/config/snapshots/$fhostname"

###################################################################################################
## DO NOT MODIFY BELOW THIS LINE OR THE SCRIPT MAY BECOME UNUSABLE!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! ##
###################################################################################################


#Set the variables stock:
curdate=`date +%Y%m%d`
host="$HOSTNAME"
host2=${host%-*}
fhostname=$host2$curdate
logfile="/opt/WCG/logs/snapshot_script.log"


#Create the log file is it does not exist already:
if [ ! -e "$logfile" ] ; then
    touch "$logfile"
fi

#Run the snapshot to the /opt/WCG/config/snapshots directory
echo "Running snapshot!"
/opt/WCG/bin/content_line -N $fhostname
echo $fhostname

#Zip up the directory for transfer:
cd $localdir
tar -czvf /tmp/$fhostname.tar.gz -C /opt/WCG/config/snapshots/$fhostname .

#Login to FTP and transfer the gzip file:
ftplog="/tmp/ftptmplog"
ftp -inv $ftpserverip <<EOF > $ftplog 
user $username $password
bin
lcd /tmp/
cd $remotedir
put $fhostname.tar.gz
quit
exit;
EOF

#Cleanup temporary files and write the success/failure to the custom log
ftpsuccess="226 Transfer complete"
if fgrep "$ftpsuccess" $ftplog ;then
   logdate=`date +%Y-%m-%d:%H:%M:%S`
   md5val=`md5sum /tmp/$fhostname.tar.gz | awk '{ print $1 }'`
   echo "$logdate - Backup $fhostname.tar.gz was created and transfered to the FTP Server! MD5 of file is: $md5val" >> $logfile
   rm -rf /opt/WCG/config/snapshots/$fhostname
   rm -rf /tmp/$fhostname.tar.gz
   rm -rf /tmp/ftptmplog
else
   logdate=`date +%Y-%m-%d:%H:%M:%S`
   echo "$logdate - Backup $fhostname.tar.gz failed, please investiagate!" >> $logfile
   rm -rf /opt/WCG/config/snapshots/$fhostname
   rm -rf /tmp/$fhostname.tar.gz
   rm -rf /tmp/ftptmplog
fi
exit 0


wcg_auto_snapshot.py
#!/usr/bin/env python
__author__ = "Alan Matson"
__copyright__ = "Copyright 2016, EchoTEK Solutions"
__license__ = "GPLv3"
__version__ = "0.94"
__maintainer__ = "Alan Matson"
__email__ = "alan (at) echotek (dot) us"
__status__ = "Development"

import socket, os, subprocess, tarfile, hashlib, time
from ftplib import FTP
from datetime import datetime
from ConfigParser import SafeConfigParser

# Import variables from config.ini and set a few others:
global server
global uname
global passw
global remoted
global locald
global currdate
global hostname
global fhostname
global logfile
global logdate
global scpstat
config = SafeConfigParser()
config.read("config.ini")
scpstat = config.get('Configuration', 'scpenable')
server = config.get('Configuration', 'serverip')
uname = config.get('Configuration', 'username')
passw = config.get('Configuration', 'password')
remoted = config.get('Configuration', 'remotedir')
locald = config.get('Configuration', 'localdir')
currdate = datetime.now().strftime('%Y%m%d')
hostname = socket.gethostname()
hostname = hostname.split("-", 1)[0]
hostname = hostname.split(".", 1)[0]
fhostname = hostname + currdate
logfile = "/opt/WCG/logs/snapshot_script.log"
logdate = str(datetime.now().strftime('%Y-%m-%d %H:%M:%S'))

def snapshot():
    #Run the snapshot on the appliance:
    #print "Running snapshot!"
    shellcmd = "/opt/WCG/bin/content_line -N %s" % fhostname
    subprocess.Popen(shellcmd, shell=True)


def tarfiles():
    #Tar the files into one compressed file:
    global tarname
    global dirname
    tarname = "/tmp/" + fhostname + ".tar.gz"
    dirname = "/opt/WCG/config/snapshots/" + fhostname
    tar = tarfile.open(tarname, 'w:gz')
    tar.add(dirname)
    tar.close()


def ftpfiles():
    #Time to FTP the file to the server:
    filen = fhostname + ".tar.gz"
    ftp = FTP(server)
    ftp.login(uname,passw)
    ftp.cwd(remoted)
    with open('/tmp/%s' % filen, 'r') as filename:
        ftpresponse = ftp.storbinary('STOR %s' % filen, filename)
    ftp.quit()

    #Check the FTP response
    if "226 Successfully transferred" not in ftpresponse:
        with open(logfile, 'a') as logf:
            logf.write("%s - Backup %s failed, please investigate!" % (logdate, filen))
    else:
        md5filen = "/tmp/" + filen
        md5sum = hashlib.md5(open(md5filen, 'rb').read()).hexdigest()
        with open(logfile, 'a') as logf:
            logf.write("%s - Backup %s was created and transfered to the FTP Server! MD5 of file is: %s" % (logdate, filen, md5sum))
    os.remove(tarname)


def scpfiles():
    filen = fhostname + ".tar.gz"
    localfile = '/tmp/' + filen
    remotefile = '/tmp/' + filen
    os.system('scp "%s" "%s@%s:%s"' % (localfile, uname, server, remotefile))


def main():
    snapshot()
    time.sleep(2)
    tarfiles()
    time.sleep(2)
    print scpstat
    if scpstat is "True" or "true":
        scpfiles()
    else:
        ftpfiles()

if __name__ == "__main__":
    main()
